File: sdk/java/src/main/java/com/anyvali/schemas/IntersectionSchema.java
package com.anyvali.schemas;

import com.anyvali.Schema;
import com.anyvali.ValidationContext;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Schema that requires input to match all given schemas.
 */
public class IntersectionSchema extends Schema {
    private final List<Schema> schemas;

    public IntersectionSchema(List<Schema> schemas) {
        this.schemas = new ArrayList<>(schemas);
    }

    private IntersectionSchema(IntersectionSchema other) {
        this.coercion = other.coercion;
        this.defaultValue = other.defaultValue;
        this.hasDefault = other.hasDefault;
        this.schemas = new ArrayList<>(other.schemas);
    }

    @Override
    protected boolean acceptsNull() {
        return schemas.stream().allMatch(Schema::acceptsNullValue);
    }

    @Override
    @SuppressWarnings("unchecked")
    protected Object validate(Object input, ValidationContext ctx) {
        List<Object> results = new ArrayList<>();
        boolean hasFailure = false;

        for (Schema schema : schemas) {
            var trialCtx = new ValidationContext(
                    new ArrayList<>(ctx.getPath()),
                    new ArrayList<>(),
                    ctx.getDefinitions()
            );
            Object result = schema.runPipeline(input, trialCtx);
            results.add(result);
            if (trialCtx.hasIssues()) {
                hasFailure = true;
                ctx.getIssues().addAll(trialCtx.getIssues());
            }
        }

        if (hasFailure) {
            return null;
        }

        // Merge map results
        if (results.stream().allMatch(r -> r instanceof Map)) {
            var merged = new LinkedHashMap<String, Object>();
            for (Object r : results) {
                merged.putAll((Map<String, Object>) r);
            }
            return merged;
        }

        // For non-map, return the last result
        return results.isEmpty() ? input : results.get(results.size() - 1);
    }

    @Override
    protected Map<String, Object> toNode() {
        var allOf = new ArrayList<Map<String, Object>>();
        for (Schema s : schemas) {
            allOf.add(s.toPortableNode());
        }
        var node = new LinkedHashMap<String, Object>();
        node.put("kind", "intersection");
        node.put("allOf", allOf);
        return addCommonNodeFields(node);
    }

    @Override
    protected Schema copy() {
        return new IntersectionSchema(this);
    }

    public List<Schema> getSchemas() {
        return schemas;
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/schemas/IntersectionSchema.kt
package com.anyvali.schemas

import com.anyvali.*
import kotlinx.serialization.json.*

data class IntersectionSchema(
    val allOf: List<Schema<*>>
) : Schema<Any?>() {
    override val kind: String = "intersection"

    override fun safeParseWithContext(input: Any?, ctx: ValidationContext): ParseResult<Any?> {
        val allIssues = mutableListOf<ValidationIssue>()
        var mergedResult: Any? = input

        for (schema in allOf) {
            val result = schema.safeParseWithContext(input, ctx)
            when (result) {
                is ParseResult.Success -> {
                    // For objects, merge results
                    if (mergedResult is Map<*, *> && result.value is Map<*, *>) {
                        @Suppress("UNCHECKED_CAST")
                        val current = (mergedResult as Map<String, Any?>).toMutableMap()
                        @Suppress("UNCHECKED_CAST")
                        current.putAll(result.value as Map<String, Any?>)
                        mergedResult = current
                    } else {
                        mergedResult = result.value
                    }
                }
                is ParseResult.Failure -> allIssues.addAll(result.issues)
            }
        }

        return if (allIssues.isEmpty()) {
            ParseResult.Success(mergedResult)
        } else {
            ParseResult.Failure(allIssues)
        }
    }

    override fun validateValue(value: Any?, ctx: ValidationContext): List<ValidationIssue> {
        val result = safeParseWithContext(value, ctx)
        return if (result is ParseResult.Failure) result.issues else emptyList()
    }

    override fun exportNode(): JsonObject = buildJsonObject {
        put("kind", JsonPrimitive("intersection"))
        put("allOf", JsonArray(allOf.map { it.exportNode() }))
    }
}
